lib/util/pthread_create_blocking_np.c

```diff
@@ -0,0 +1,170 @@
+#include <sys/time.h>
+
+#include <errno.h>
+#include <pthread.h>
+#include <stdlib.h>
+
+#include "pthread_create_blocking_np.h"
+
+struct wrapped_cookie {
+	/* User-supplied. */
+	void * (*start_routine)(void *);
+	void * arg;
+
+	/* Synchronization. */
+	pthread_mutex_t mutex;
+	pthread_cond_t cond;
+	int running;		/* 1 if the thread has started. */
+
+	/* In case the synchronization fails. */
+	int rc_sync;		/* non-zero if synchronization failed. */
+};
+
+/* Set the "running" flag and signal the condition variable. */
+static int
+set_running(struct wrapped_cookie * U)
+{
+	int rc;
+
+	/* Lock mutex. */
+	if ((rc = pthread_mutex_lock(&U->mutex)))
+		goto err0;
+
+	/* Set the "running" flag and signal the condition variable. */
+	U->running = 1;
+	if ((rc = pthread_cond_signal(&U->cond)))
+		goto err1;
+
+	/* Unlock mutex. */
+	if ((rc = pthread_mutex_unlock(&U->mutex)))
+		goto err0;
+
+	/* Success! */
+	return (0);
+
+err1:
+	/*
+	 * Don't record any errors in this clean-up; the existing value of rc
+	 * is more important.
+	 */
+	pthread_mutex_unlock(&U->mutex);
+err0:
+	/* Failure! */
+	return (rc);
+}
+
+/* Routine which is executed by pthread_create(). */
+static void *
+wrapped_thread(void * cookie)
+{
+	struct wrapped_cookie * U = cookie;
+	void * (*start_routine)(void *);
+	void * arg;
+	int rc;
+
+	/* Copy the user-supplied parameters, because U will not be valid. */
+	start_routine = U->start_routine;
+	arg = U->arg;
+
+	/* Indicate that the main thread can continue (which will free U). */
+	if ((rc = set_running(U)) != 0) {
```

Comment:
I would merge `set_running` into here.  I don't see any benefit to having it as a separate function.

Reply:
I thought it would be easier to reason about?  I mean, `set_running` is the bog-standard way to signal a condition variable.  I figured that if that was a separate function, it would be easy to check, then the reader could progress to `wrapped_thread`, which would also be easier to understand due to the signalling and mutexes not being part of it.

I've merged it as you requested, but I put that as the final commit in this branch, in case you take a look at the merged code and reconsider.